.github/workflows/lib_compat_test.py

```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env python3
+
+# Verifies a binary uses only dynamic symbols from whitelisted library versions.
+# Prints the disallowed symbols and their versions on failure.
+# Usage: lib_compat_test.py bin/clangd --lib=GLIBC_2.18
+
+import argparse
+import subprocess
+import sys
+
+parser = argparse.ArgumentParser()
+parser.add_argument('binary')
+parser.add_argument('--lib', action='append', default=[],
+                    help="Whitelist a library, e.g. GLIBC_2.18 or GLIBC")
+parser.add_argument('--sym', action='append', default=[],
+                    help="Whitelist a symbol, e.g. crc32")
+args = parser.parse_args()
+
+# Parses GLIBC_2.3 into ("GLIBC", [2,3])
+# Parses GLIBC into ("GLIBC", None)
+def parse_version(version):
+  parts = version.split('_', 1)
+  if len(parts) == 1:
+    return (version, None)
+  return (parts[0], [int(p) for p in parts[1].split('.')])
+
+lib_versions = dict([parse_version(v) for v in args.lib])
+
+# Determines whether an optionally-versioned symbol is acceptable.
+def accept_symbol(sym):
+  if sym in args.sym:
+    return True
+  split = sym.split('@', 1)
+  return (split[0] in args.sym) or (len(split) == 2 and accept_lib(split[1]))
+
+# Determines whether all symbols with version 'lib' are acceptable.
+def accept_lib(lib):
+  (lib, ver) = parse_version(lib)
+  if not lib in lib_versions: # Non-whitelisted library.
+    return False
+  if lib_versions[lib] is None: # Unrestricted version
+    return True
+  if ver is None: # Lib has non-numeric version, library restricts version.
+    return False
+  return ver <= lib_versions[lib]
+
+# Run nm to find the undefined symbols, and check whether each is acceptable.
+nm = subprocess.run(['nm', '-uD', '--with-symbol-version', args.binary],
+                    stdout=subprocess.PIPE, text=True)
+nm.check_returncode()
+status = 0
+for line in nm.stdout.splitlines():
+  # line = "       U foo@GLIBC_2.3"
+  parts = line.split()
+  if len(parts) != 2:
+    print("Unparseable nm output: ", line, file=sys.stderr)
+    status = 2
+    continue
+  if parts[0] == 'w': # Weak-undefined symbol, not actually required.
+    continue
+  if not accept_symbol(parts[1]):
+    print(parts[1])
```

Comment:
maybe some info before the symbol name?

`found unsuitable symbol:`

Reply:
The symbol names are the only output, apart from the message on stderr explaining what they are :-)
This makes it easier to use this in a shell pipeline without having to `cut` all the time...